vllm/model_executor/models/mixtral_quant.py

```diff
@@ -129,31 +132,66 @@ def __init__(
     def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         num_tokens, hidden_dim = hidden_states.shape
         hidden_states = hidden_states.view(-1, hidden_dim)
-        # router_logits: (num_tokens, n_experts)
         router_logits, _ = self.gate(hidden_states)
 
-        routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
-        routing_weights, selected_experts = torch.topk(routing_weights,
-                                                       self.top_k,
-                                                       dim=-1)
-        routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
-
-        final_hidden_states = None
-        for expert_idx in self.expert_indicies:
-            expert_layer = self.experts[expert_idx]
-            expert_mask = (selected_experts == expert_idx)
-            expert_weights = (routing_weights * expert_mask).sum(dim=-1,
-                                                                 keepdim=True)
-
-            current_hidden_states = expert_layer(hidden_states).mul_(
-                expert_weights)
-            if final_hidden_states is None:
-                final_hidden_states = current_hidden_states
-            else:
-                final_hidden_states.add_(current_hidden_states)
+        qweight13_l = []
+        scales13_l = []
+        qweight2_l = []
+        scales2_l = []
+
+        for i in range(len(self.experts)):
+            w1_qw = self.experts[i].w1.get_parameter("qweight").int()
+            w3_qw = self.experts[i].w3.get_parameter("qweight").int()
+            w1_s = self.experts[i].w1.get_parameter("scales").half()
+            w3_s = self.experts[i].w3.get_parameter("scales").half()
+            w2_qw = self.experts[i].w2.get_parameter("qweight").int()
+            w2_s = self.experts[i].w2.get_parameter("scales").half()
```

Comment:
are these guaranteed to be fp16?

Reply:
From Pytorch documentation: `self.half() is equivalent to self.to(torch.float16)`
Scales are not necessarily fp16 when loaded.